### PR TITLE
my remarks

### PR DESCRIPTION
**Pull request by blainepwnz**
Hello!
My name is Andrew.
I have been one who interviewed you yesterday.
Here are some remarks about your test task.
Feel free to ask questions if you have some.
Have a nice day :)

### CODE REVIEW

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java [hunk 1]
```diff
@@ -73,17 +72,14 @@ public static void start(Context context, int position) {
...
-        if (savedInstanceState != null) {
-            position = savedInstanceState.getInt(IMAGE_POSITION_KEY);
-        }else  {
-            position = getIntent().getIntExtra(IMAGE_POSITION_KEY, 0);
-        }
         setContentView(R.layout.activity_main);
         ButterKnife.bind(this);
         setSupportActionBar(toolbar);
         imageList = new ArrayList<>();
+        //is it necessary to clear empty list?
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:79]: I agree it isn't necessary

File: TitaniumTask/app/src/main/java/com/sa/testtask/api/Response.java [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 public class Response {
 
+    //if name of field in same as in serialized string, it can be omitted
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/api/Response.java:9]: nice! I didnt know it. Thanks

File: TitaniumTask/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -23,6 +23,7 @@
         </activity>
         <activity
             android:name=".screens.MainActivity"
+            android:launchMode="singleTask"
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/AndroidManifest.xml:26]: I have known about it, but didnt use correct. If I now right understood this mode will call onNewIntent with new parameter  "int position" in Intent?

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/ListActivity.java [hunk 4]
```diff
@@ -54,18 +52,31 @@ protected void onCreate(Bundle savedInstanceState) {
         recyclerView.setLayoutManager(new LinearLayoutManager(this));
         recyclerView.setAdapter(adapter);
         showList(imageList);
-        Subscribe();
+        //should subscribe in resume
+        //if on stop will be called all subscriptions will be lost
+        //Subscribe();
+    }
 
+    @Override
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/ListActivity.java:60]: I agree and I know about it, but forgot to change (in MainActivity it is changed)

File: TitaniumTask/app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (blainepwnz) [reply TitaniumTask/app/src/main/AndroidManifest.xml:26]: If you have existing activity all future intents will proceed to OnNewIntent.

You had another option:
1) Start list activity for result
2) In list activity on select of image setResult(intent with image position)
3) Finish list activity
4) Get intent with position in main activity

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java [hunk 5]
```diff
@@ -61,7 +58,9 @@
 
     @OnTextChanged(R.id.edit_label)
     void onEditTextClick(CharSequence s, int start, int before, int count) {
-        if (s.length() != count) save.setVisibility(View.VISIBLE);
+        //it is not working
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:61]: Really? It should work. If you add or delete one character, button save will become visible.

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java [hunk 6]
```diff
@@ -125,35 +114,47 @@ private void update(int position) {
...
+    protected void onNewIntent(Intent intent) {
+        super.onNewIntent(intent);
+        //just for example
+        Log.d(TAG, "onNewIntent: got new intent");
+        position = intent.getIntExtra(IMAGE_POSITION_KEY, 0);
+        update(position);
     }
 
     private void startListActivity() {
+        //you are creating multiple instances of this activity every time user selecting image
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:155]: I think you are right, it really better solution. I havent thought about onActivityResult ((((

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java (hunk 5 above)
Comment (blainepwnz) [reply TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:61]: Tested on nexus 5x with android 8, problem was that count == s.length all the time.
Only when i pressed enter on IME it helped.

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/SplashActivity.java [hunk 7]
```diff
@@ -38,24 +38,24 @@ protected void onResume() {
...
-                .flatMap(integer -> Api.getImagesApi().getImages())
-                .filter(response -> response != null)
-                .map(Response::getImages)
-                .filter(images -> images != null && images.size() > 0)
-                .doOnNext(images -> Storage.getInstance().setImages(images))
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(images -> startMainActivity(),
-                        throwable -> Log.d(TAG, throwable.getMessage(), throwable));
+            //should wrap as exeption or do it in doOnNext
+            .doOnNext(isConnected -> {
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/SplashActivity.java:42]: ok thanks your version is more clear

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/SplashActivity.java [hunk 8]
```diff
@@ -65,6 +65,7 @@ private void startMainActivity() {
     }
 
     private void showProgressDialog() {
+        //bad practice from android 8
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/SplashActivity.java:68]: Yes I know, for now it should be progressBar

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/list/ImagesRVAdapter.java [hunk 9]
```diff
@@ -34,20 +35,27 @@ public void swapData(List<Response.Image> imageList) {
...
+    // very complicated
+    public PublishSubject<Integer> getViewPositionClickedSubject() {
+        return publishPositionSubject;
+    }
+
     @Override
     public ImagesVH onCreateViewHolder(ViewGroup parent, int viewType) {
         LayoutInflater inflater = LayoutInflater.from(parent.getContext());
         View view = inflater.inflate(R.layout.item_image, parent, false);
+        //used only to create view, attaching click listeners here is a bad practice
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/list/ImagesRVAdapter.java:48]: I agree! should be changed))

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java [hunk 10]
```diff
@@ -97,13 +93,6 @@ private void Subscribe() {
         subscription.add(onSaveClick());
         subscription.add(onRandomClick());
         subscription.add(onSelectClick());
-
-    }
-
-    @Override
```
Comment (aSvetenco) [inline TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:103]: why is it not necessary? If screen orientation will change Activity will be destroyed and will show image that came when Activity created the first time

File: TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java (hunk 5 above)
Comment (aSvetenco) [reply TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java:61]: Interesting)) I also tested on 5x android 8 (real device) and also on htc with android 5 and it was working. But ok I will try to do somehow another

**Comment by aSvetenco** [comment]
I did some fixes. Thank you (Andrew) a lot for your remarks and explanation.

**Pull request closed without merging**